README.md
Evertrue Fork of the Kibana Library Cookbook
===============

Original available [here](https://github.com/lusis/chef-kibana)

A library cookbook for installing kibana 4.

If you wish to install Kibana3 try using the '1.3.x' releases of the cookbook.

This is a library cookbook,  you can use you included `kibana_lwrp::install` recipe to install kibana, but it is more likely to be really useful for you to call the LWRPs directly from your own cookbook.  You can use the `kibana_lwrp::install` recipe as a guide and customize to your needs.

This cookbook is tested and works well with the `logstash` and the `elasticsearch` cookbooks found on the chef community site.

If you'd like to point kibana to a cluster of Elasticsearch servers instead of a single node, please review the most current recommended instructions at:
 https://www.elastic.co/guide/en/kibana/current/production.html#load-balancing

Attributes
----------
As with most cookbooks I write, this one is hopefully flexible enough to be wrapped by allowing you to override as much as possible

See `attributes/*.rb` for attributes ( documented inline )

Usage
-----
#### kibana_lwrp::default

The default recipe does nothing except for allow access to the LWRPs.

#### kibana_lwrp::install

This is designed to be an example recipe to show you how you might write a wrapper cookbook.   However it should be usable for a
simple install of kibana.

- install kibana4 `/opt/kibana/XXXX` and create a symlink called `current` in the same directory.
- install `nginx` and serve the kibana application

If you wish to swap `apache` for `nginx`, simply set `node['kibana']['webserver']` to `apache` in a role/environment/node somewhere.

If you don't want this cookbook to handle the webserver config simply set `node['kibana']['webserver']` to `''` in a role/environment/node somewhere.
Please note that in this case you have to set `node['kibana']['user']`.

Both cookbooks, by default, will configure the appropriate proxy to your ElasticSearch server such that you don't have to expose it to the world.

**NOTE**
There is **NO** security enabled by default on any of the content being served.
If you would like to modify the `nginx` or `apache` parameters, you should:

- create your own cookbook i.e. `my-kibana`
- copy the template for the webserver you wish to use to your cookbook
- modify the template as you see fit (add auth, setup ssl)
- use the appropriate webserver template attributes to point to your cookbook and template

#### kibana::examples

examples of using the LWRPs.  Don't use this directly.

Testing
-------
#### kitchen


```
$ bundle install
$ bundle exec berks install
$ bundle exec rake style spec # spec and style checks
$ bundle exec rake kitchen # integration testing
```


Contributing
------------
- Fork the repository on Github
- Create a named feature branch (like `add_component_x`)
- Write you change
- Write tests for your change (if applicable)
- Run the tests, ensuring they all pass
-- `bundle exec strainer test`
- Submit a Pull Request using Github

License and Authors
-------------------
Primary authors:

- John E. Vincent <lusis.org+github.com@gmail.com>
- Paul Czarkowski <username.taken@gmail.com>

Contributors:

- Jeff Hubbard (@lord2800)
- @mouadino
- Ben P (@benwtr)
- Chris Ferry (@chrisferry)
- Ian Neubert (@ianneub)
- kellam (@klamontagne)
- Martin Smith (@martinb3)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# CHANGELOG for et_kibana_lwrp

This file is used to list changes made in each version of kibana.

## 1.1.0

* Switch to "releases" changelog

## 1.0.1

* Fork!
* Switch to community-maintained **chef_nginx** cookbook

# Pre-Evertrue fork

## 3.0.2
* Fix a notification that was to runit_service[nginx] instead of service[nginx].

## 3.0.1
* Fix a leftover poise relic -- default_action
* Remove notifying_block that remained

## 3.0.0
* Fix a number of Rubocop and Foodcritic complaints
* Update README.md for contributors, documentation (#26)
* Don't break kibana.yml template when values are nil (#111)
* Warn about URL and Version attributes when used incorrectly (#107)
* Remove poise, add `provides` DSL, and update test fixture to ES 2.1.0
* Add kibana_port to apache kibana web resource
* Make default_app_id configurable via attribute (#101)
* Remove duplicate slashes in paths (#102)
* Bump kibana file method to 4.2.1
* Begin testing with Ubuntu 14.04 instead of 12.04

## 2.0.6
* fix apache template

## 2.0.4
* fix tests
* support kibana 4 official

## 2.0.2
* revert use inline resources

## 2.0.0
* change cookbook name to kibana_lwrp
* support kibana 4 beta

## 1.4.1
* bugfix - could not update kibana because of remote_file name

## 1.4.0
* initial support for kibana4
* removed support for kibana3 ( use 1.3.x releases )

## 1.3.2:
* mostly bugfixes

## 1.3.0:
* big rewrite to act as library cookbook.
* not sure if it will be backwards compat yet.

## 1.2.1:
* proxy backend healthcheck
* update proxy templates to match kibana master
* changed config.js template to pull from attributes

## 1.2.0:

* Add ability to choose install type of either git repo or zipfile
* Add web_dir attribute to handle kibana location for different install types
* Add attributes for http scheme for both webserver and es, helpful if they're managed outside of this cookbook.
* Some Attribute namespace cleanup,  read through carefully if upgrading.
?????? * Update config.js to newer format

## 1.1.5

* Added Strainer test suite
* Allow installing kibana without a webserver
* Boolean Attribute kibana['git']['checkout'] to allow default git behavior to be checkout rather than sync.


## 1.1.4

* Update configuration to use proxy_pass setup by cookbook

## 1.1.3

* Rename dist to src symlink as per latest kibana folder structure change

## 1.1.2

* Update config.js.erb to be the same as the new config.js source from kibana

## 1.1.1

* Fixed previous symlink of whole repository, now it only symlinks the dist directory

## 1.1.0

* Changed paths in default recipe for kibana source to the new kibana directory structure

## 1.0.0:

* Initial release of kibana


test/fixtures/kibana_test/metadata.rb
name 'kibana_test'
maintainer 'John E. Vincent'
maintainer_email 'lusis.org+github.com@gmail.com'
license 'Apache 2.0'
description 'Installs/Configures kibana_test'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))

version '0.1.0'

depends 'et_kibana_lwrp'
depends 'elasticsearch'


test/fixtures/kibana_test/recipes/default.rb

include_recipe 'elasticsearch::default'
include_recipe 'et_kibana_lwrp::install'

elasticsearch_service 'elasticsearch' do
  action :start
end


libraries/web.rb
# Encoding: utf-8
# Cookbook Name:: kibana
# Resource/Provider:: web
# Author:: John E. Vincent
# Author:: Paul Czarkowski
# License:: Apache 2.0
#
# Copyright 2014, John E. Vincent

class Chef
  class Resource::KibanaWeb < ChefCompat::Resource
    provides :kibana_web
    default_action(:create)

    property(:name, String, name_property: true)
    property(:type, String, default: 'nginx')
    property(:template_cookbook, String, default: 'et_kibana_lwrp')
    property(:template, String)
    property(:server_name, String, default: 'kibana.localhost')
    property(:server_aliases, Array, default: ['kibana'])
    property(:listen_address, String, default: '0.0.0.0')
    property(:listen_port, String, default: '80')
    property(:es_server, String, default: '127.0.0.1')
    property(:es_port, String, default: '9200')
    property(:es_scheme, String, default: 'http://')
    property(:version, String, default: lazy { node['kibana']['version'] })
    property(:default_site_enabled, [true, false], default: false)
    property(:docroot, String, default: '/opt/kibana/current/kibana')
    property(:kibana_port, Integer, default: 5601)
  end

  class Provider::KibanaWeb < Chef::Provider::LWRPBase # ~FC057, ~FC058
    provides :kibana_web
    include Chef::DSL::Recipe # required under chef 12, see poise/poise #8

    action :create do
      converge_by("create resource #{new_resource.name}") do
        resources = kibana_resources

        case resources[:type]
        when 'apache'
          node.set['apache']['default_site_enabled'] = resources[:default_site_enabled]

          node.set['apache']['listen_ports'] = [] unless resources[:default_site_enabled]
          node.set['apache']['listen_ports'] << resources[:listen_port]

          %w(apache2 apache2::mod_dir apache2::mod_proxy apache2::mod_proxy_http).each do |recipe|
            @run_context.include_recipe recipe
          end

          wa = web_app resources[:name] do
            cookbook resources[:template_cookbook]
            docroot resources[:docroot]
            template resources[:template]
            es_server resources[:es_server]
            es_port resources[:es_port]
            server_name resources[:server_name]
            server_aliases resources[:server_aliases]
            kibana_dir resources[:docroot]
            listen_address resources[:listen_address]
            listen_port resources[:listen_port]
            es_scheme resources[:es_scheme]
            kibana_port resources[:kibana_port]
          end

        when 'nginx'
          node.set['nginx']['default_site_enabled'] = resources[:default_site_enabled]
          node.set['nginx']['install_method'] = node['kibana']['nginx']['install_method']
          @run_context.include_recipe 'chef_nginx'

          template "#{node['nginx']['dir']}/sites-available/#{resources[:name]}" do
            source resources[:template]
            cookbook resources[:template_cookbook]
            notifies :reload, 'service[nginx]'
            variables(
              es_server: resources[:es_server],
              es_port: resources[:es_port],
              server_name: resources[:server_name],
              server_aliases: resources[:server_aliases],
              kibana_dir: resources[:docroot],
              listen_address: resources[:listen_address],
              listen_port: resources[:listen_port],
              es_scheme: resources[:es_scheme],
              kibana_port: resources[:kibana_port]
            )
          end
          nginx_site resources[:name]
        when ''
          # do nothing
        else
          Chef::Application.fatal!("Unknown type: #{resources[:type]}")
        end # end case
      end # end converge by
    end # end def

    action :remove do
      converge_by("remove resource #{new_resource.name}") do
        # Normal Chef recipe code goes here
      end
    end

    private

    def kibana_resources
      kb = {
        name: new_resource.name,
        type: new_resource.type,
        template_cookbook: new_resource.template_cookbook,
        docroot: new_resource.docroot,
        template: new_resource.template || "kibana-#{new_resource.type}.conf.erb",
        server_name: new_resource.server_name,
        server_aliases: new_resource.server_aliases,
        listen_address: new_resource.listen_address,
        listen_port: new_resource.listen_port,
        es_server: new_resource.es_server,
        es_port: new_resource.es_port,
        es_scheme: new_resource.es_scheme,
        default_site_enabled: new_resource.default_site_enabled,
        version: new_resource.version,
        kibana_port: new_resource.kibana_port
      }
      kb
    end
  end
end


recipes/install.rb
#
# Cookbook Name:: kibana
# Recipe:: install
#
# Copyright 2013, John E. Vincent
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

install_type = node['kibana']['install_type']

unless Chef::Config[:solo]
  es_server_results = search(:node, "roles:#{node['kibana']['es_role']} AND chef_environment:#{node.chef_environment}")
  unless es_server_results.empty?
    node.set['kibana']['es_server'] = es_server_results[0]['ipaddress']
  end
end

if node['kibana']['user'].empty?
  if !node['kibana']['webserver'].empty?
    webserver = node['kibana']['webserver']
    kibana_user = node[webserver]['user']
  else
    kibana_user = 'nobody'
  end
else
  kibana_user = node['kibana']['user']
  kibana_user kibana_user do
    name kibana_user
    group kibana_user
    home node['kibana']['install_dir']
    action :create
  end
end

kibana_install 'kibana' do
  user kibana_user
  group kibana_user
  install_dir node['kibana']['install_dir']
  install_type install_type
  action :create
end

docroot = "#{node['kibana']['install_dir']}/current/kibana"
kibana_config = "#{node['kibana']['install_dir']}/current/#{node['kibana'][install_type]['config']}"
es_server = "#{node['kibana']['es_scheme']}#{node['kibana']['es_server']}:#{node['kibana']['es_port']}"

template kibana_config do
  source node['kibana'][install_type]['config_template']
  cookbook node['kibana'][install_type]['config_template_cookbook']
  mode '0644'
  user kibana_user
  group kibana_user
  variables(
    index: node['kibana']['config']['kibana_index'],
    port: node['kibana']['java_webserver_port'],
    elasticsearch: es_server,
    default_route: node['kibana']['config']['default_route'],
    panel_names:  node['kibana']['config']['panel_names'],
    default_app_id: node['kibana']['config']['default_app_id']
  )
end

if install_type == 'file'

  include_recipe 'java::default' if node['kibana']['install_java']
  include_recipe 'runit::default'

  runit_service 'kibana' do
    options(
      user: kibana_user,
      home: "#{node['kibana']['install_dir']}/current"
    )
    cookbook 'et_kibana_lwrp'
    subscribes :restart, "template[#{kibana_config}]", :delayed
  end

end

kibana_web 'kibana' do
  type lazy { node['kibana']['webserver'] }
  docroot docroot
  es_server node['kibana']['es_server']
  kibana_port node['kibana']['java_webserver_port']
  template 'kibana-nginx_file.conf.erb'
  not_if { node['kibana']['webserver'] == '' }
end

include_recipe 'et_kibana_lwrp::plugins'


attributes/default.rb
cookbook_name = 'et_kibana_lwrp'

# The method used to install kibana.  `git` will clone the git repo,
# `file` will download from elasticsearch.org
# git is not really supported since the move to java based server.
default['kibana']['install_type'] = 'file' # git | file
default['kibana']['version'] = '4.6.2-linux-x86_64' # must match version number of kibana being installed

# Values to use for git method of installation
default['kibana']['git']['url'] = 'https://github.com/elasticsearch/kibana'
default['kibana']['git']['branch'] = 'v4.6.2'
default['kibana']['git']['type'] = 'sync' # checkout | sync
default['kibana']['git']['config'] = 'kibana/config.js' # relative path of config file
default['kibana']['git']['config_template'] = 'config.js.erb' # template to use for config
default['kibana']['git']['config_template_cookbook'] = cookbook_name # cookbook containing config template

# Values to use for file method of installation
default['kibana']['file']['type'] = 'tgz' # zip | tgz

default['kibana']['file']['url'] = nil # calculated based on version, unless you override this
default['kibana']['file']['checksum'] = nil # sha256 ( shasum -a 256 FILENAME )
default['kibana']['file']['config'] = 'config/kibana.yml' # relative path of config file
default['kibana']['file']['config_template'] = 'kibana.yml.erb' # template to use for config
default['kibana']['file']['config_template_cookbook'] = cookbook_name # cookbook containing config template

# Kibana Java Web Server
default['kibana']['java_webserver_port'] = 5601

# this is only used by the recipe.  if you use the LWRPs
# (which you should) then install java from your own recipe.
default['kibana']['install_java'] = true

# Which webserver to use, and webserver options.
default['kibana']['webserver'] = 'nginx' # nginx or apache
default['kibana']['webserver_hostname'] = node.name
default['kibana']['webserver_aliases'] = [node['ipaddress']]
default['kibana']['webserver_listen'] = node['ipaddress']
default['kibana']['webserver_port'] = 80
default['kibana']['webserver_scheme'] = 'http://'

# parent directory of install_dir.  This is required because of the `file` method.
default['kibana']['install_path'] = '/opt'

# the actual installation directory of kibana. If using the `file` method this should be left as is.
default['kibana']['install_dir'] = "#{node['kibana']['install_path']}/kibana"

# used to configure proxy information for the webserver to proxy ES calls.
default['kibana']['es_server'] = '127.0.0.1'
default['kibana']['es_port'] = '9200'
default['kibana']['es_role'] = 'elasticsearch_server'
default['kibana']['es_scheme'] = 'http://'

# user to install kibana files as.  if left blank will use the default webserver user.
default['kibana']['user'] = 'kibana'

# config template location and variables.
default['kibana']['config']['kibana_index']  = 'kibana-int'
default['kibana']['config']['panel_names']   = %w(histogram map pie table filtering timepicker text fields hits dashcontrol column derivequeries trends bettermap query terms)
default['kibana']['config']['default_route'] = '/dashboard/file/logstash.json'
default['kibana']['config']['default_app_id'] = 'discover'
# include quote inside this next variable if not using window.location style variables...
# e.g.  = "'http://elasticsearch.example.com:9200'"
default['kibana']['config']['elasticsearch'] = 'window.location.protocol+"//"+window.location.hostname+":"+window.location.port'

# nginx variables
default['kibana']['nginx']['install_method'] = 'package'
default['kibana']['nginx']['template'] = 'kibana-nginx.conf.erb'
default['kibana']['nginx']['template_cookbook'] = cookbook_name
default['kibana']['nginx']['enable_default_site'] = false
default['kibana']['nginx']['install_method'] = 'package'

# Apache variables.
default['kibana']['apache']['template'] = 'kibana-apache.conf.erb'
default['kibana']['apache']['template_cookbook'] = cookbook_name
default['kibana']['apache']['enable_default_site'] = false
